src/core/seat/input-method-relay.cpp

```diff
@@ -0,0 +1,293 @@
+        {
+            assert(text_input->input->focused_surface == nullptr);
+            if (surface != text_input->pending_focused_surface)
+            {
+                text_input->set_pending_focused_surface(nullptr);
+            }
+        } else
+        if (text_input->input->focused_surface != nullptr)
+        {
+            assert(text_input->pending_focused_surface == nullptr);
+            if (surface != text_input->input->focused_surface)
+            {
+                disable_text_input(text_input->input);
+                wlr_text_input_v3_send_leave(text_input->input);
+            } else
+            {
+                LOGD("set_focus an already focused surface");
+                continue;
+            }
+        }
+
+        if (surface && wl_resource_get_client(text_input->input->resource)
+            == wl_resource_get_client(surface->resource))
+        {
+            if (input_method)
+            {
+                wlr_text_input_v3_send_enter(text_input->input, surface);
+            } else
+            {
+                text_input->set_pending_focused_surface(surface);
+            }
+        }
+    }
+}
+
+wf_input_method_relay::~wf_input_method_relay()
+{
+}
+
+wf_text_input::wf_text_input(wf_input_method_relay *rel, wlr_text_input_v3 *in) :
+    relay(rel), input(in), pending_focused_surface(nullptr)
+{
+
+    on_text_input_enable.set_callback([&] (void *data)
+    {
+        auto wlr_text_input = static_cast<wlr_text_input_v3*> (data);
+        assert(input == wlr_text_input);
+
+        if (relay->input_method == nullptr)
+        {
+            LOGI("Enabling text input, but input method is gone");
+            return;
+        }
+
+        wlr_input_method_v2_send_activate(relay->input_method);
+        relay->send_im_state(input);
+    });
+
+    on_text_input_commit.set_callback([&] (void *data)
+    {
+        auto wlr_text_input = static_cast<wlr_text_input_v3*> (data);
+        assert(input == wlr_text_input);
+
+        if (!input->current_enabled)
+        {
+            LOGI("Inactive text input tried to commit");
+            return;
+        }
+
+        if (relay->input_method == nullptr)
+        {
+            LOGI("Committing text input, but input method is gone");
+            return;
+        }
+
+        relay->send_im_state(input);
+    });
+
+    on_text_input_disable.set_callback([&] (void *data)
+    {
+        auto wlr_text_input = static_cast<wlr_text_input_v3*> (data);
+        assert(input == wlr_text_input);
+
+        relay->disable_text_input(input);
+    });
+
+    on_text_input_destroy.set_callback([&] (void *data)
+    {
+        auto wlr_text_input = static_cast<wlr_text_input_v3*> (data);
+        assert(input == wlr_text_input);
+
+        if (input->current_enabled)
+        {
+            relay->disable_text_input(input);
+        }
+        set_pending_focused_surface(nullptr);
+        on_text_input_enable.disconnect();
```

Comment:
Are these the disconnects which you needed to not get a crash?

Seems to me that the object will be immediately deleted so they should be auto-disconnected.